src/main/java/net/snowflake/client/core/HttpUtil.java

```diff
@@ -69,16 +68,53 @@
 
   public static final String JDBC_TTL = "net.snowflake.jdbc.ttl";
 
-  /** The unique httpClient shared by all connections. This will benefit long- lived clients */
-  private static Map<OCSPMode, CloseableHttpClient> httpClient = new ConcurrentHashMap<>();
+  /**
+   * Object to provide the proxy user and password credentials for the HTTP Client object when
+   * needed
+   */
+  private static CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+
+  private static SnowflakeProxyRoutePlanner sdkProxyRoutePlanner = null;
 
   /**
-   * The unique httpClient shared by all connections that don't want decompression. This will
-   * benefit long-lived clients
+   * The unique httpClient shared by all connections. This will benefit long- lived clients. Key =
+   * proxy host + proxy port + nonProxyHosts, Value = Map<OCSPMode, HttpClient>
    */
-  private static Map<OCSPMode, CloseableHttpClient> httpClientWithoutDecompression =
+  private static Map<String, Map<OCSPMode, CloseableHttpClient>> httpClientWithProxies =
       new ConcurrentHashMap<>();
 
+  /**
+   * The unique httpClient map shared by all connections that don't want decompression. This will
+   * benefit long-lived clients. Key = proxy host + proxy port + nonProxyHosts, Value =
+   * Map<OCSPMode, HttpClient>
+   */
+  private static Map<String, Map<OCSPMode, CloseableHttpClient>>
```

Comment:
consider to create a class called httpClientOptions as the key which includes OCSPMode and other proxy settings.